pkg/controller/support/credentials/cloudauth/gcr/gcr_test.go

```
@@ -1,102 +1,321 @@
+	}{
+		{
+			"invalid-url",
+			"test-ts",
+			defaultCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: nil,
+			},
+			nil,
+			defaultChallengeLoginServer,
+			invalidUrlErr.Error(),
+			invalidUrlErr,
+		},
+		{
+			"oauth2-error",
+			"gcr.io",
+			defaultCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{errOut: true},
+			},
+			nil,
+			defaultChallengeLoginServer,
+			gcrTokenAccessErr.Error(),
+			gcrTokenAccessErr,
+		},
+		{
+			"failed-challenge-ts-error",
+			"hi-docker.pkg.dev",
+			defaultCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				"realmName",
+				defaultTestingErr),
+			gcrRegistryErr.Error(),
+			gcrRegistryErr,
+		},
+		{
+			"failed-do-request",
+			"gcr.io",
+			canceledCtx,
+			nil,
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			ctxCanceledErr.Error(),
+			ctxCanceledErr,
+		},
+		{
+			"non-200-response-code",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				return &http.Response{
+					StatusCode: http.StatusBadRequest,
+					Body:       io.NopCloser(strings.NewReader(`nope`)),
+				}, nil
+			}),
+			&gcrProvider{
+				logger:      log,
+				tokenSource: &fakeOauth2TokenSource{},
+			},
+			nil,
+			cloudauthtest.FakeChallengeLoginServer(
+				"serviceName",
+				ts.URL,
+				nil),
+			non200StatusErr.Error(),
+			non200StatusErr,
+		},
+		{
+			"failed-no-token-in-response",
+			"gcr.io",
+			defaultCtx,
+			roundTripFunc(func(r *http.Request) (*http.Response, error) {
+				res, err := json.Marshal(tokenResponse{})
 				if err != nil {
-					t.Fatalf("%#v", err)
+					t.Fatalf("Unexpected error while marshalling token response")
```

don't swallow the actual error?